README.md
# Material Color Properties

This package streamlines the process of accessing properties from the Material Design 3 theme.

> [!WARNING]  
> This repository is used for personal purposes.
> It is not recommended to use this package in production.
> 

## Quick Setup

Add `@webhead/material-color-properties` dependency to your project, also add `@material/material-color-utilities` if  you haven't already.

```bash
# Using pnpm
pnpm i @webhead/material-color-properties

# Using yarn
yarn add  @webhead/material-color-properties

# Using npm
npm install @webhead/material-color-properties
```

## Example

### Getting the theme via the `themeFromSourceColor` function
```ts
import {argbFromHex, CustomColor, Theme, themeFromSourceColor} from "@material/material-color-utilities";

const theme: Theme = themeFromSourceColor(argbFromHex('#ff0088'), [
    {
        name: 'Arab Green',
        value: '#22a45a',
        blend: true,
    }
])
```

### Getting the properties via the `propertiesFromTheme` function
```ts
import {propertiesFromTheme} from '@webhead/material-color-properties'

// Default
const options = {
   tones: [0, 5, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
   dark: false,
   brightnessSuffix: true,
   prefix: {
      palette: 'md-ref-palette-',
      color: 'md-sys-color-',
      customColor: 'md-custom-color-',
   }
}

const properties = propertiesFromTheme(theme, options) 
```

### Example Output

```
{
    // Palette Primary
    "--md-ref-palette-primary0": "#000000",
    "--md-ref-palette-primary5": "#111111",
    "--md-ref-palette-primary10": "#1b1b1b",
    "--md-ref-palette-primary20": "#303030",
    "--md-ref-palette-primary30": "#474747",
    "--md-ref-palette-primary40": "#5e5e5e",
    "--md-ref-palette-primary50": "#777777",
    "--md-ref-palette-primary60": "#919191",
    "--md-ref-palette-primary70": "#ababab",
    "--md-ref-palette-primary80": "#c6c6c6",
    "--md-ref-palette-primary95": "#f1f1f1",
    "--md-ref-palette-primary99": "#fcfcfc",
    "--md-ref-palette-primary100": "#ffffff",
        
    // Palette Secondary
    "--md-ref-palette-secondary0": "#000000",
    "--md-ref-palette-secondary5": "#111111",
    "--md-ref-palette-secondary10": "#1b1b1b",
    "--md-ref-palette-secondary20": "#303030",
    "--md-ref-palette-secondary30": "#474747",
    "--md-ref-palette-secondary40": "#5e5e5e",
    "--md-ref-palette-secondary50": "#777777",
    "--md-ref-palette-secondary60": "#919191",
    "--md-ref-palette-secondary70": "#ababab",
    "--md-ref-palette-secondary80": "#c6c6c6",
    "--md-ref-palette-secondary95": "#f1f1f1",
    "--md-ref-palette-secondary99": "#fcfcfc",
    "--md-ref-palette-secondary100": "#ffffff",
        
    // Palette Tertiary
    "--md-ref-palette-tertiary0": "#000000",
    "--md-ref-palette-tertiary5": "#111111",
    "--md-ref-palette-tertiary10": "#1b1b1b",
    "--md-ref-palette-tertiary20": "#303030",
    "--md-ref-palette-tertiary30": "#474747",
    "--md-ref-palette-tertiary40": "#5e5e5e",
    "--md-ref-palette-tertiary50": "#777777",
    "--md-ref-palette-tertiary60": "#919191",
    "--md-ref-palette-tertiary70": "#ababab",
    "--md-ref-palette-tertiary80": "#c6c6c6",
    "--md-ref-palette-tertiary95": "#f1f1f1",
    "--md-ref-palette-tertiary99": "#fcfcfc",
    "--md-ref-palette-tertiary100": "#ffffff",
        
    // Palette Neutral
    "--md-ref-palette-neutral0": "#000000",
    "--md-ref-palette-neutral5": "#111111",
    "--md-ref-palette-neutral10": "#1b1b1b",
    "--md-ref-palette-neutral20": "#303030",
    "--md-ref-palette-neutral30": "#474747",
    "--md-ref-palette-neutral40": "#5e5e5e",
    "--md-ref-palette-neutral50": "#777777",
    "--md-ref-palette-neutral60": "#919191",
    "--md-ref-palette-neutral70": "#ababab",
    "--md-ref-palette-neutral80": "#c6c6c6",
    "--md-ref-palette-neutral95": "#f1f1f1",
    "--md-ref-palette-neutral99": "#fcfcfc",
    "--md-ref-palette-neutral100": "#ffffff",
        
    // Palette Neutral Variant
    "--md-ref-palette-neutral-variant0": "#000000",
    "--md-ref-palette-neutral-variant5": "#111111",
    "--md-ref-palette-neutral-variant10": "#1b1b1b",
    "--md-ref-palette-neutral-variant20": "#303030",
    "--md-ref-palette-neutral-variant30": "#474747",
    "--md-ref-palette-neutral-variant40": "#5e5e5e",
    "--md-ref-palette-neutral-variant50": "#777777",
    "--md-ref-palette-neutral-variant60": "#919191",
    "--md-ref-palette-neutral-variant70": "#ababab",
    "--md-ref-palette-neutral-variant80": "#c6c6c6",
    "--md-ref-palette-neutral-variant95": "#f1f1f1",
    "--md-ref-palette-neutral-variant99": "#fcfcfc",
    "--md-ref-palette-neutral-variant100": "#ffffff",
        
    // Palette Error
    "--md-ref-palette-error0": "#000000",
    "--md-ref-palette-error5": "#111111",
    "--md-ref-palette-error10": "#1b1b1b",
    "--md-ref-palette-error20": "#303030",
    "--md-ref-palette-error30": "#474747",
    "--md-ref-palette-error40": "#5e5e5e",
    "--md-ref-palette-error50": "#777777",
    "--md-ref-palette-error60": "#919191",
    "--md-ref-palette-error70": "#ababab",
    "--md-ref-palette-error80": "#c6c6c6",
    "--md-ref-palette-error95": "#f1f1f1",
    "--md-ref-palette-error99": "#fcfcfc",
    "--md-ref-palette-error100": "#ffffff",
   
    // Scheme
    "--md-sys-color-primary": "#ffb1c7",
    "--md-sys-color-on-primary": "#650032",
    "--md-sys-color-primary-container": "#8e0049",
    "--md-sys-color-on-primary-container": "#ffd9e2",
    "--md-sys-color-secondary": "#e3bdc6",
    "--md-sys-color-on-secondary": "#422931",
    "--md-sys-color-secondary-container": "#5b3f47",
    "--md-sys-color-on-secondary-container": "#ffd9e2",
    "--md-sys-color-tertiary": "#efbd93",
    "--md-sys-color-on-tertiary": "#47290b",
    "--md-sys-color-tertiary-container": "#613f1f",
    "--md-sys-color-on-tertiary-container": "#ffdcc1",
    "--md-sys-color-error": "#ffb4ab",
    "--md-sys-color-on-error": "#690005",
    "--md-sys-color-error-container": "#93000a",
    "--md-sys-color-on-error-container": "#ffdad6",
    "--md-sys-color-background": "#201a1b",
    "--md-sys-color-on-background": "#ece0e1",
    "--md-sys-color-surface": "#201a1b",
    "--md-sys-color-on-surface": "#ece0e1",
    "--md-sys-color-surface-variant": "#514346",
    "--md-sys-color-on-surface-variant": "#d5c2c5",
    "--md-sys-color-outline": "#9e8c90",
    "--md-sys-color-outline-variant": "#514346",
    "--md-sys-color-shadow": "#000000",
    "--md-sys-color-scrim": "#000000",
    "--md-sys-color-inverse-surface": "#ece0e1",
    "--md-sys-color-inverse-on-surface": "#352f30",
    "--md-sys-color-inverse-primary": "#ba0061",
   
    // Scheme Light
    "--md-sys-color-primary-light": "#ba0061",
    "--md-sys-color-on-primary-light": "#ffffff",
    "--md-sys-color-primary-container-light": "#ffd9e2",
    "--md-sys-color-on-primary-container-light": "#3e001d",
    "--md-sys-color-secondary-light": "#74565e",
    "--md-sys-color-on-secondary-light": "#ffffff",
    "--md-sys-color-secondary-container-light": "#ffd9e2",
    "--md-sys-color-on-secondary-container-light": "#2b151c",
    "--md-sys-color-tertiary-light": "#7c5635",
    "--md-sys-color-on-tertiary-light": "#ffffff",
    "--md-sys-color-tertiary-container-light": "#ffdcc1",
    "--md-sys-color-on-tertiary-container-light": "#2e1500",
    "--md-sys-color-error-light": "#ba1a1a",
    "--md-sys-color-on-error-light": "#ffffff",
    "--md-sys-color-error-container-light": "#ffdad6",
    "--md-sys-color-on-error-container-light": "#410002",
    "--md-sys-color-background-light": "#fffbff",
    "--md-sys-color-on-background-light": "#201a1b",
    "--md-sys-color-surface-light": "#fffbff",
    "--md-sys-color-on-surface-light": "#201a1b",
    "--md-sys-color-surface-variant-light": "#f2dde1",
    "--md-sys-color-on-surface-variant-light": "#514346",
    "--md-sys-color-outline-light": "#837376",
    "--md-sys-color-outline-variant-light": "#d5c2c5",
    "--md-sys-color-shadow-light": "#000000",
    "--md-sys-color-scrim-light": "#000000",
    "--md-sys-color-inverse-surface-light": "#352f30",
    "--md-sys-color-inverse-on-surface-light": "#faeeef",
    "--md-sys-color-inverse-primary-light": "#ffb1c7",
    
    // Scheme Dark
    "--md-sys-color-primary-dark": "#ba0061",
    "--md-sys-color-on-primary-dark": "#ffffff",
    "--md-sys-color-primary-container-dark": "#ffd9e2",
    "--md-sys-color-on-primary-container-dark": "#3e001d",
    "--md-sys-color-secondary-dark": "#74565e",
    "--md-sys-color-on-secondary-dark": "#ffffff",
    "--md-sys-color-secondary-container-dark": "#ffd9e2",
    "--md-sys-color-on-secondary-container-dark": "#2b151c",
    "--md-sys-color-tertiary-dark": "#7c5635",
    "--md-sys-color-on-tertiary-dark": "#ffffff",
    "--md-sys-color-tertiary-container-dark": "#ffdcc1",
    "--md-sys-color-on-tertiary-container-dark": "#2e1500",
    "--md-sys-color-error-dark": "#ba1a1a",
    "--md-sys-color-on-error-dark": "#ffffff",
    "--md-sys-color-error-container-dark": "#ffdad6",
    "--md-sys-color-on-error-container-dark": "#410002",
    "--md-sys-color-background-dark": "#fffbff",
    "--md-sys-color-on-background-dark": "#201a1b",
    "--md-sys-color-surface-dark": "#fffbff",
    "--md-sys-color-on-surface-dark": "#201a1b",
    "--md-sys-color-surface-variant-dark": "#f2dde1",
    "--md-sys-color-on-surface-variant-dark": "#514346",
    "--md-sys-color-outline-dark": "#837376",
    "--md-sys-color-outline-variant-dark": "#d5c2c5",
    "--md-sys-color-shadow-dark": "#000000",
    "--md-sys-color-scrim-dark": "#000000",
    "--md-sys-color-inverse-surface-dark": "#352f30",
    "--md-sys-color-inverse-on-surface-dark": "#faeeef",
    "--md-sys-color-inverse-primary-dark": "#ffb1c7",
   
    // Surface Level
    "--md-sys-color-surface-level0": "#ff008800",
    "--md-sys-color-surface-level1": "#ff00880d",
    "--md-sys-color-surface-level2": "#ff008814",
    "--md-sys-color-surface-level3": "#ff00881c",
    "--md-sys-color-surface-level4": "#ff00881f",
    "--md-sys-color-surface-level5": "#ff008824",
   
    // Surface Level Light
    "--md-sys-color-surface-level0-light": "#ff008800",
    "--md-sys-color-surface-level1-light": "#ff00880d",
    "--md-sys-color-surface-level2-light": "#ff008814",
    "--md-sys-color-surface-level3-light": "#ff00881c",
    "--md-sys-color-surface-level4-light": "#ff00881f",
    "--md-sys-color-surface-level5-light": "#ff008824",
    
    // Surface Level Dark
    "--md-sys-color-surface-level0-dark": "#ff008800",
    "--md-sys-color-surface-level1-dark": "#ff00880d",
    "--md-sys-color-surface-level2-dark": "#ff008814",
    "--md-sys-color-surface-level3-dark": "#ff00881c",
    "--md-sys-color-surface-level4-dark": "#ff00881f",
    "--md-sys-color-surface-level5-dark": "#ff008824",
    
    // Custom Color Palette
    "--md-ref-palette-arab-green0": "#000000",
    "--md-ref-palette-arab-green5": "#00150d",
    "--md-ref-palette-arab-green10": "#002116",
    "--md-ref-palette-arab-green20": "#003828",
    "--md-ref-palette-arab-green30": "#00513b",
    "--md-ref-palette-arab-green40": "#006c50",
    "--md-ref-palette-arab-green50": "#008865",
    "--md-ref-palette-arab-green60": "#05a47b",
    "--md-ref-palette-arab-green70": "#3cc095",
    "--md-ref-palette-arab-green80": "#5ddcaf",
    "--md-ref-palette-arab-green95": "#bcffe1",
    "--md-ref-palette-arab-green99": "#f4fff7",
    "--md-ref-palette-arab-green100": "#ffffff",
        
    // Custom Color Scheme
    "--md-custom-color-arab-green": "#8ed96d",
    "--md-custom-color-on-arab-green": "#0e3900",
    "--md-custom-color-arab-green-container": "#185200",
    "--md-custom-color-on-arab-green-container": "#a9f686",
        
    // Custom Color Scheme Light
    "--md-custom-color-arab-green-light": "#276c0a",
    "--md-custom-color-on-arab-green-light": "#ffffff",
    "--md-custom-color-arab-green-container-light": "#a9f686",
    "--md-custom-color-on-arab-green-container-light": "#052100",
   
    // Custom Color Scheme Dark   
    "--md-custom-color-arab-green-dark": "#276c0a",
    "--md-custom-color-on-arab-green-dark": "#ffffff",
    "--md-custom-color-arab-green-container-dark": "#a9f686",
    "--md-custom-color-on-arab-green-container-dark": "#052100",
     
    // Scheme RGB Values
    "--md-sys-color-primary-rgb": "255, 177, 199",
    "--md-sys-color-on-primary-rgb": "101, 0, 50",
    "--md-sys-color-primary-container-rgb": "142, 0, 73",
    "--md-sys-color-on-primary-container-rgb": "255, 217, 226",
    "--md-sys-color-secondary-rgb": "227, 189, 198",
    "--md-sys-color-on-secondary-rgb": "66, 41, 49",
    "--md-sys-color-secondary-container-rgb": "91, 63, 71",
    "--md-sys-color-on-secondary-container-rgb": "255, 217, 226",
    "--md-sys-color-tertiary-rgb": "239, 189, 147",
    "--md-sys-color-on-tertiary-rgb": "71, 41, 11",
    "--md-sys-color-tertiary-container-rgb": "97, 63, 31",
    "--md-sys-color-on-tertiary-container-rgb": "255, 220, 193",
    "--md-sys-color-error-rgb": "255, 180, 171",
    "--md-sys-color-on-error-rgb": "105, 0, 5",
    "--md-sys-color-error-container-rgb": "147, 0, 10",
    "--md-sys-color-on-error-container-rgb": "255, 218, 214",
    "--md-sys-color-background-rgb": "32, 26, 27",
    "--md-sys-color-on-background-rgb": "236, 224, 225",
    "--md-sys-color-surface-rgb": "32, 26, 27",
    "--md-sys-color-on-surface-rgb": "236, 224, 225",
    "--md-sys-color-surface-variant-rgb": "81, 67, 70",
    "--md-sys-color-on-surface-variant-rgb": "213, 194, 197",
    "--md-sys-color-outline-rgb": "158, 140, 144",
    "--md-sys-color-outline-variant-rgb": "81, 67, 70",
    "--md-sys-color-shadow-rgb": "0, 0, 0",
    "--md-sys-color-scrim-rgb": "0, 0, 0",
    "--md-sys-color-inverse-surface-rgb": "236, 224, 225",
    "--md-sys-color-inverse-on-surface-rgb": "53, 47, 48",
    "--md-sys-color-inverse-primary-rgb": "186, 0, 97",
        
    // Scheme Light RGB Values
    "--md-sys-color-primary-light-rgb": "186, 0, 97",
    "--md-sys-color-on-primary-light-rgb": "255, 255, 255",
    "--md-sys-color-primary-container-light-rgb": "255, 217, 226",
    "--md-sys-color-on-primary-container-light-rgb": "62, 0, 29",
    "--md-sys-color-secondary-light-rgb": "116, 86, 94",
    "--md-sys-color-on-secondary-light-rgb": "255, 255, 255",
    "--md-sys-color-secondary-container-light-rgb": "255, 217, 226",
    "--md-sys-color-on-secondary-container-light-rgb": "43, 21, 28",
    "--md-sys-color-tertiary-light-rgb": "124, 86, 53",
    "--md-sys-color-on-tertiary-light-rgb": "255, 255, 255",
    "--md-sys-color-tertiary-container-light-rgb": "255, 220, 193",
    "--md-sys-color-on-tertiary-container-light-rgb": "46, 21, 0",
    "--md-sys-color-error-light-rgb": "186, 26, 26",
    "--md-sys-color-on-error-light-rgb": "255, 255, 255",
    "--md-sys-color-error-container-light-rgb": "255, 218, 214",
    "--md-sys-color-on-error-container-light-rgb": "65, 0, 2",
    "--md-sys-color-background-light-rgb": "255, 251, 255",
    "--md-sys-color-on-background-light-rgb": "32, 26, 27",

CHANGELOG.md
# @webhead/material-color-properties

## 2.3.1

### Patch Changes

- 66a6cec: Hotfix

## 2.3.0

### Minor Changes

- c67b031: Upgrade of dependencies to latest

## 2.2.0

### Minor Changes

- 9b245e9: Added properties for the new surface color roles

## 2.1.6

### Patch Changes

- cb4394e: fixed issue with tokenize fn

## 2.1.5

### Patch Changes

- 30926aa: feat: added option for rgb formatting

## 2.1.4

### Patch Changes

- a1c8d15: chore: removed package as strict peerDependency and changed internal usage back to @material

## 2.1.3

### Patch Changes

- 331828f: Fix of wrong formatting for custom palette color property

## 2.1.2

### Patch Changes

- 47800d2: fix of wrong import
- 3bc6188: revert of change to @material
- c9d3c06: revert of change to @material

## 2.1.1

### Patch Changes

- d3df6d9: chore: added comments

## 2.1.0

### Minor Changes

- 9f5da87: Added tests for main functionality

## 2.0.2

### Patch Changes

- 7d6c9c8: changes

## 2.0.1

### Patch Changes

- dcf89b8: fix

## 2.0.0

### Major Changes

- 807c1fd: fix of mising arguments of propertiesFromTheme fn

### Patch Changes

- b31f71c: chore: replaced @material with @importantimport as peerDependency, bc bug in original package built process

## 1.0.4

### Patch Changes

- 3b15830: refactor: fix README.md, arguments to main method

## 1.0.3

### Patch Changes

- 6c71033: Moved stuff to /utils and changed README.md

## 1.0.2

### Patch Changes

- 8093501: Fix in readme.md and main method

## 1.0.1

### Patch Changes

- e615f67: Updated README.md and fixed issue with propertiesFromTheme

## 1.0.0

### Major Changes

- 7400220: Correctly works now, did change fn name

## 0.0.7

### Patch Changes

- 5461684: fix

## 0.0.6

### Patch Changes

- 0eebfd0: trying things

## 0.0.5

### Patch Changes

- fa23634: dependency fix
- 4eac0a3: another fix

## 0.0.4

### Patch Changes

- d665a2c: version patch

## 0.0.3

### Patch Changes

- aabe2c5: fix for no dist folder
- 189dc00: fix for missing dist folder

## 0.0.2

### Patch Changes

- 0d35300: first changeset


src/index.ts
import {
    argbFromHex,
    Blend,
    CustomColorGroup,
    hexFromArgb,
    Scheme,
    Theme, themeFromSourceColor,
    TonalPalette
} from "@material/material-color-utilities";

import {hexAFromArgb, rgbFromHex, tokenize} from "./utils";
import chalk from "chalk";
import {log} from "util";

/**
 * Defines a configuration object that can
 * have optional properties for a color palette.
 * @typedef {Object} ConfigOptions
 */
interface ConfigOptions {
    // Whether to append a suffix to color names that indicate their brightness level.
    brightnessSuffix?: boolean,
    // Whether the configuration is for a dark mode.
    dark?: boolean,
    // An array of numbers representing the tones to be used in the color palette.
    tones?: number[],
    // An object that can have optional properties for the name of the palette,
    // the name of the base color, and a custom color.
    prefix?: {
        palette?: string,
        color?: string,
        customColor?: string,
    },
    // rgb options config
    rgb?: {
        // whether to include rgb values
        include?: boolean, // default is true
        // whether to include alpha values
        separator?: ',' | ' ', // default is ','
        // name of the suffix
        suffix?: string, // default is -rgb
    }
}

/**
 * A utility type that defines the properties of a CSS custom property.
 * @typedef {Object<string, string>} Properties
 */
type Properties = Record<string, string>

/**
 * A utility type that recursively makes all properties of an object required.
 * @typedef {Object} RequiredDeep
 */
type RequiredDeep<T> = {
    [P in keyof T]-?: RequiredDeep<T[P]>
}

/**
 * A type that is derived from `ConfigOptions` and makes all its properties required.
 * @typedef {Object} DefaultConfig
 */
type DefaultConfig = RequiredDeep<ConfigOptions>

/**
 * @param {Theme} theme - The theme to derive custom properties from
 * @param {Object} options - An object containing optional properties for customizing the derived properties
 * @param {Array<number>} options.tones - An array of numbers representing the tones to derive custom properties for
 * @param {string} [options.prefix] - An optional prefix to add to the beginning of each derived custom property name
 * @returns {Properties} An object with CSS custom properties derived from the theme
 */
function derivePaletteProperties(theme: Theme, {
    tones,
    prefix
}: { tones: number[], prefix?: string }): Properties {
    const properties = {} as Record<string, string>
    for (const [key, value] of Object.entries(theme.palettes)) {
        const tonalPalette = value instanceof TonalPalette ? value : TonalPalette.fromInt(value)
        for (const tone of tones) {
            properties[`--${prefix ?? ''}${tokenize(key)}${tone}`] = hexFromArgb(tonalPalette.tone(tone))
        }
    }
    return properties
}

/**
 * @param scheme The color scheme to derive properties from.
 * @param prefix An optional prefix to add to each custom property.
 * @param suffix An optional suffix to add to each custom property.
 * @returns An object with CSS custom properties derived from the color scheme.
 */
function deriveSchemeProperties(scheme: Scheme, {
    prefix,
    suffix
}: { prefix?: string, suffix?: string }): Properties {
    const properties = {} as Record<string, string>
    const entries = Object.entries(typeof scheme?.toJSON === 'function' ? scheme.toJSON() : scheme)
    for (const [key, value] of entries) {
        properties[`--${prefix ?? ''}${tokenize(key)}${suffix ?? ''}`] = hexFromArgb(value)
    }
    return properties
}

/**
 * @param argb The color to derive properties from.
 * @param options An object with optional properties for a prefix, suffix, and tones.
 */
function deriveSurfaceElevationProperties(argb: number, options: {
    prefix?: string,
    suffix?: string,
    tones?: number[]
}) {
    const properties = {} as Record<string, string>
    const tones = options?.tones ?? [0, 0.05, 0.08, 0.11, 0.12, 0.14]
    tones.forEach((tone, index) => {
        const color = hexAFromArgb(argb, tone)
        const token = `--${options?.prefix ?? ''}surface-level${index}${options?.suffix ?? ''}`
        properties[token] = color
    })
    return properties
}

/**
 * @param {Record<string, string>} properties - A Record object containing CSS custom properties with hex color values.
 * @param separator
 * @param suffix
 * @returns {Record<string, string>} A new Record object with CSS custom properties and their corresponding RGB color values.
 */
function rgbProperties(properties: Record<string, string>, {
    separator,
    suffix
}: { separator: ',' | ' ', suffix: string } = {
    separator: ',',
    suffix: '-rgb'
}) {
    return Object.entries(properties).reduce((acc, [name, hex]) => ({
        ...acc,
        [`${name}${suffix}`]: rgbFromHex(hex, separator)
    }), {} as Record<string, string>)
}

function paletteFromCustomColor(customColor: CustomColorGroup, {
    tones,
    sourceColor
}: { tones: number[], sourceColor: number }) {
    let tonalPalette = TonalPalette.fromInt(customColor.color.value)
    if (customColor.color.blend) {
        const harmonized = Blend.harmonize(customColor.color.value, sourceColor)
        tonalPalette = TonalPalette.fromInt(harmonized)
    }
    tones.forEach(tone => tonalPalette.tone(tone))
    return tonalPalette
}

function deriveCustomSchemeProperties(colorGroup: CustomColorGroup, options?: {
    suffix?: string,
    prefix?: string,
    dark?: boolean
}) {
    const properties = {} as Record<string, string>
    for (const [key, value] of Object.entries(colorGroup[options?.dark ? 'dark' : 'light'])) {
        const name = tokenize(colorGroup.color.name)
        const map = key.replace(/([a-z])(?=[A-Z])/g, '$1-').toLowerCase()
        const token = map.replace(/color/, name)
        properties[`--${options?.prefix ?? ''}${token}${options?.suffix ?? ''}`] = hexFromArgb(value as number)
    }
    return properties
}

function deriveCustomPaletteProperties(customColors: CustomColorGroup[], {
    tones,
    prefix,
    suffix,
    sourceColor
}: { tones: number[], prefix?: string, suffix?: string, sourceColor: number }) {
    const properties: Record<string, string> = {}
    for (const customColor of customColors) {
        const palette = paletteFromCustomColor(customColor, {tones, sourceColor})
        tones.forEach((tone: number) => {
            const name = tokenize(customColor.color.name)
            const token = `--${prefix ?? ''}${name}${tone}${suffix ?? ''}`
            properties[token] = hexFromArgb(palette.tone(tone))
        })
    }
    return properties
}

const getSurfaceContainers = (
    sourceColor: number,
    options?: {
        dark?: boolean,
        prefix?: string,
        suffix?: string
    },
) => {
    const roleToneMap = {
        'surface-dim': {
            light: 87,
            dark: 6,
        },
        'surface-bright': {
            light: 98,
            dark: 24
        },
        'surface-container-lowest': {
            light: 100,
            dark: 4
        },
        'surface-container-low': {
            light: 96,
            dark: 10
        },
        'surface-container': {
            light: 94,
            dark: 12
        },
        'surface-container-high': {
            light: 92,
            dark: 17
        },
        'surface-container-highest': {
            light: 90,
            dark: 22
        },
    }
    const palette = TonalPalette.fromInt(sourceColor)
    return Object.entries(roleToneMap).reduce((acc, [role, tone]) => {
        const tonalColor = palette.tone(tone[options?.dark ? 'dark' : 'light'])
        const name = `--${options?.prefix ?? 'md-sys-color-'}${role}${options?.suffix ?? ''}`
        return {
            ...acc,
            [name]: hexFromArgb(tonalColor)
        }
    }, {} as Record<string, string>)
}

const getSurfaceTints = (
    surfaceColor: number,
    primaryColor: number,
    options?: {
        suffix?: string,
        prefix?: string
    }
) => {
    const schemeNames = [
        'background-tint'
    ]
    const prefix = options?.prefix ?? 'md-sys-color-'
    const suffix = options?.suffix ?? ''
    const tints = [0.05, 0.08, 0.11, 0.12, 0.14];
    return tints.reduce((colors, tint, index) => {
        const color = Blend.cam16Ucs(surfaceColor, primaryColor, tint)
        const name = schemeNames[index]
        colors[`--${prefix}${name}${suffix}`] = hexFromArgb(color)
        return colors
    }, {} as Record<string, string>)
};


/**
 * @param {Theme} theme - A Theme object.
 * @param {object} options - An options object containing the tones and prefix properties.
 * @param {number[]} options.tones - An array of tones to apply to the palette.
 * @param {string} options.prefix - A prefix to prepend to each CSS custom property name.
 * @returns {Record<string, string>} - Object containing CSS custom property names and values.
 */
function deriveProperties(theme: Theme, {
    dark,
    tones,
    brightnessSuffix,
    prefix,
    rgb
}: {
    dark: boolean,
    tones: number[],
    brightnessSuffix?: boolean,
    prefix: {
        palette: string,
        color: string,
        customColor: string
    },
    rgb: {
        include: boolean,
        separator: ',' | ' '
        suffix: string
    }
}) {

    const palettes = derivePaletteProperties(theme, {
        tones,
        prefix: prefix?.palette
    })

    const schemes = {
        ...deriveSchemeProperties(theme.schemes[dark ? 'dark' : 'light'], {
            prefix: prefix?.color
        }),
        ...brightnessSuffix
            ? {
                ...deriveSchemeProperties(theme.schemes.light, {
                    prefix: prefix?.color,
                    suffix: '-light'
                }),
                ...deriveSchemeProperties(theme.schemes.dark, {
                    prefix: prefix?.color,
                    suffix: '-dark'
                })
            }
            : {}
    }

    const ccPalettes = deriveCustomPaletteProperties(theme.customColors, {
        tones,
        prefix: prefix?.palette,
        sourceColor: theme.source
    })

    const ccSchemes = theme.customColors.reduce((acc, color) => ({
        ...acc,
        ...deriveCustomSchemeProperties(color, {
            prefix: prefix?.customColor,
            dark
        }),
        ...brightnessSuffix
            ? {
                ...deriveCustomSchemeProperties(color, {
                    prefix: prefix?.customColor,
                    suffix: '-light'
                }),
                ...deriveCustomSchemeProperties(color, {
                    prefix: prefix?.customColor,
                    suffix: '-dark',
                    dark: true
                })
            }
            : {}
    }), {})


    // const surfaceElevations = {
    //     ...deriveSurfaceElevationProperties(dark
    //             ? theme.schemes.dark.primary
    //             : theme.schemes.light.primary,
    //         {prefix: prefix?.color}
    //     ),
    //     ...brightnessSuffix
    //         ? {
    //             ...deriveSurfaceElevationProperties(theme.schemes.light.primary, {
    //                 prefix: prefix?.color,
    //                 suffix: '-light'
    //             }),
    //             ...deriveSurfaceElevationProperties(theme.schemes.dark.primary, {
    //                 prefix: prefix?.color,
    //                 suffix: '-dark',
    //             })
    //         }
    //         : {}
    // }
    
    const surfaceElevations = {
        ...deriveSurfaceElevationProperties(dark
                ? theme.schemes.dark.primary
                : theme.schemes.light.primary,
            {prefix: prefix?.color}
        ),
        ...brightnessSuffix
            ? {
                ...deriveSurfaceElevationProperties(theme.schemes.light.primary, {
                    prefix: prefix?.color,
                    suffix: '-light'
                }),
                ...deriveSurfaceElevationProperties(theme.schemes.dark.primary, {
                    prefix: prefix?.color,
                    suffix: '-dark',
                })
            }
            : {}
    }

    const surfaceContainers = {
        ...getSurfaceContainers(
            theme.schemes[dark ? 'dark' : 'light'].surface,
            {dark, prefix: prefix?.color}
        ),
        ...brightnessSuffix
            ? {
                ...getSurfaceContainers(
                    theme.schemes.light.surface,
                    {suffix: '-light', prefix: prefix?.color}
                ),
                ...getSurfaceContainers(
                    theme.schemes.dark.surface,
                    {dark: true, suffix: '-dark', prefix: prefix?.color}
                )
            }
            : {}
    }

    const palettesRgb = rgbProperties(palettes, {
        separator: rgb.separator,
        suffix: rgb.suffix
    })

    const schemesRgb = rgbProperties(schemes, {
        separator: rgb.separator,
        suffix: rgb.suffix
    })

    const ccSchemesRgb = rgbProperties(ccSchemes, {
        separator: rgb.separator,
        suffix: rgb.suffix
    })

    const surfaceContainersRgb = rgbProperties(surfaceContainers, {
        separator: rgb.separator,
        suffix: rgb.suffix
    })

    return {
        ...palettes,
        ...schemes,
        ...ccPalettes,
        ...ccSchemes,
        ...surfaceElevations,
        ...surfaceContainers,
        ...({...rgb.include && surfaceContainersRgb}),
        ...({...rgb.include && palettesRgb}),
        ...({...rgb.include && schemesRgb}),
        ...({...rgb.include && ccSchemesRgb}),
    }
}


/**
 * @param {Theme} theme - A Theme object.
 * @param {object} [options] - Optional configuration for prefix, suffix, and dark mode.
 * @param {string} [options.prefix] - Optional prefix to prepend to each CSS custom property name.
 * @param {string} [options.suffix] - Optional suffix to append to each CSS custom property name.
 * @param {boolean} [options.dark] - Optional boolean flag indicating whether to use the "dark" color scheme.
 * @returns {Record<string, string>} - Object containing CSS custom property names and values.
 * @example
 * const theme = themeFromSourceColor(argbFromHex('#ff0000'))
 * // => { source: 4294901760, schemes: { light: { ... }, dark: { ... } } }
 * const properties = propertiesFromTheme(theme)
 * // => { '--md-ref-palette-primary100': '#ffffff', '--md-sys-color-primary': '#ff0000', ... }
 */
const propertiesFromTheme = (theme: Theme, options?: ConfigOptions): Properties => {
    const defaultConfig: DefaultConfig = {
        tones: [0, 5, 10, 20, 30, 40, 50, 60, 70, +80, 90, 100],
        dark: false,
        brightnessSuffix: true,
        prefix: {
            palette: 'md-ref-palette-',
            color: 'md-sys-color-',
            customColor: 'md-custom-color-',
        },
        rgb: {
            include: true,
            separator: ' ',
            suffix: '-rgb'
        }
    }
    return deriveProperties(theme, {
        ...defaultConfig,
        ...options,
        prefix: {
            ...defaultConfig.prefix,
            ...(options?.prefix ?? {}),
        },
        rgb: {
            ...defaultConfig.rgb,
            ...(options?.rgb ?? {})
        }
    })
}

export {
    propertiesFromTheme,
    hexAFromArgb,
    rgbFromHex,
    tokenize,
}